Support for hierarchical collections

DigitalGoldfish opened the pull request:
As discussed in https://github.com/fikaproductions/fika-gatsby-source-cockpit/issues/11 I added support for hierarchical organized collections.

It works as far as I can tell but I am not totally sure if I use the best way to add the child nodes to the GraphQL tree. Looking at the [Gatsby Node APIs](https://www.gatsbyjs.org/docs/node-creation/) documentation I can see that there are methods to link nodes in a parent/child relationship however when I tried to implement this all i got in the Query result in the "children" field was an array of string ids, not the object.

DigitalGoldfish commented:
One more thing that came to my mind just now ... would it be meaningful to add a field "depth" or "level" to the collection nodes? I am not sure if it is a useless addition or might make sense for some use cases ...

WhippetsAintDogs commented:
I think that adding a "depth" field might be useless since the GraphQL query will output the same hierarchy of nested objects/entries as represented in Cockpit and we're not flattening by any means the whole structure.

DigitalGoldfish commented:
The entries are already returned flat for an "All" query - since we only ask for entries of a given type and all of them share the same type this is standard graphQL. 

Given my data structure from #11 a graphQL to allCockpitMenu would return all the menu items regardless of hierarchy in a flat edges array since the child items have the same type as the parent item and that's all we ask from GraphQL here. 

After thinking about it a bit longer I should at least add a link to the parent item to the children to make it possible to query all children without getting their parent too. Still not sure about the depth field ... it would make it possible to query all items of depth x regardless of parent but I am not sure how common a use case that is ... personally I can't think of any right now for me but of course others might disagree ....

WhippetsAintDogs commented:
All right, I agree with, that makes sens. 

Regarding your first comment, I just remembered a reason why you shouldn't use parent/children node relationship and continue your implementation like you are doing with foreign keys: these are made for transformer plugin, children are supposed to be derivated data from the parent and there is the whole notion of cascade node deletion with parent/children node links (if we delete the parent, it deletes the children also). I'm not 100% sure, but I tend to think that what you did is more appropriate for this use case.

DigitalGoldfish commented:
I added a link from the child to the parent by adding a field "parent" that links to the parent node. I also added another attribute "level" that contains the hierarchy level of the current node. 

One slightly weird thing I noticed is that the children are not only in the "children" field but also in the "cockpitChildren" field. I have no explanation why that happens (doesn't seem to be in the plugin) so I assume that this is some Gatsby magic. Makes me wonder though if i should use cockpitChildren/cockpitParent as field names to avoid any possible conflicts.

WhippetsAintDogs commented:
@DigitalGoldfish, the answer for the "cockpitChildren/cockpitParent" is here under the API > Inputs section: https://github.com/angeloashmore/gatsby-node-helpers. 

You can do like I do in the MarkdownNodeFactory by deleting the field afterward if it holds the same value that its non-"cockpit" prefixed counter part. ☺

DigitalGoldfish commented on src/CollectionItemNodeFactory.js:
```diff
@@ -15,22 +15,31 @@ module.exports = class CollectionItemNodeFactory {
+
+    const node = nodeFactory(collectionItem);
+    this.createNode(node);
+    return node;
```
Do you have a linting configuration that you can share for this project? Or maybe even add to the repository.

DigitalGoldfish commented on src/CollectionItemNodeFactory.js:
```diff
@@ -15,22 +15,31 @@ module.exports = class CollectionItemNodeFactory {
+        return this.create(childItem);
+      })
+      : [];
+
```
Thanks for the explanation - finally it makes sense why I had the additional field for children but not for the parent.

WhippetsAintDogs commented:
fix #11
